[Style] Replace isZero/isNegative/isPositive functions inherited from WebCore::Length

weinig opened the pull request:
#### 33b130ec17ce2cc298ac28ac3a693e14a271d863
<pre>
[Style] Replace isZero/isNegative/isPositive functions inherited from WebCore::Length
<a href="https://bugs.webkit.org/show_bug.cgi?id=299972">https://bugs.webkit.org/show_bug.cgi?id=299972</a>

Reviewed by Darin Adler.

`Style::LengthPercentage&lt;&gt;` and `Style::LengthWrapperBase&lt;&gt;` inherited (to help
things continue to work without too much churn) the isZero/isNegative/isPositive
functions from `WebCore::Length` and along with them their quirky behavior with
calc and keywords.

For a type that includes a Length, Percentage, Calc, and Keywords, the
functions have the following behavior:

- isZero:
    Length        !value
    Percentage    !value
    Calc          false
    Keyword       true (due to keywords having a value equal to 0)

- isPositive:
    Length        value &gt; 0
    Percentage    value &gt; 0
    Calc          true
    Keyword       false (due to keywords having a value equal to 0)

- isNegative:
    Length        value &lt; 0
    Percentage    value &lt; 0
    Calc          false
    Keyword       false (due to keywords having a value equal to 0)

This change replaces these functions with a set that hopefully have
more clear semantics:

The &quot;known&quot; value functions, `isKnownZero`, `isKnownPositive` and
`isKnownNegative`, will perform their comparison for Length and
Percentage values, but will return `false` for anything else (e.g. calc
or keywords).

The &quot;possibly&quot; value functions, `isPossiblyZero`, `isPossiblyPositive` and
`isPossiblyNegative`, will perform their comparison for Length and
Percentage values, but will return `true` for anything else (e.g. calc
or keywords).

In addition, the functions are constrained only to types where the types
numeric range makes the comparison meaningful. This means that a type
like `LengthPercentage&lt;CSS::Nonnegative&gt;`, the `isKnownNegative` and
`isPossiblyNegative` functions will unavailable and produce a compile
error if used. This allowed the removal of a number of useless calls.

* Source/WebCore/css/values/CSSValueConcepts.h:
* Source/WebCore/layout/formattingContexts/FormattingGeometry.cpp:
* Source/WebCore/layout/formattingContexts/block/BlockMarginCollapse.cpp:
* Source/WebCore/rendering/AutoTableLayout.cpp:
* Source/WebCore/rendering/GridLayoutFunctions.cpp:
* Source/WebCore/rendering/RenderBlock.cpp:
* Source/WebCore/rendering/RenderBlockFlow.cpp:
* Source/WebCore/rendering/RenderBox.cpp:
* Source/WebCore/rendering/RenderElement.cpp:
* Source/WebCore/rendering/RenderImage.cpp:
* Source/WebCore/rendering/RenderMarquee.cpp:
* Source/WebCore/rendering/RenderTable.cpp:
* Source/WebCore/rendering/RenderTableSection.cpp:
* Source/WebCore/rendering/cocoa/RenderThemeCocoa.mm:
* Source/WebCore/rendering/style/BorderData.cpp:
* Source/WebCore/rendering/style/BorderData.h:
* Source/WebCore/rendering/style/RenderStyle.cpp:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
* Source/WebCore/style/StyleExtractorCustom.h:
* Source/WebCore/style/values/StyleValueTypes.h:
* Source/WebCore/style/values/backgrounds/StyleBackgroundSize.h:
* Source/WebCore/style/values/backgrounds/StyleFillLayers.h:
* Source/WebCore/style/values/box/StyleMargin.h:
* Source/WebCore/style/values/fill-stroke/StyleStrokeWidth.cpp:
* Source/WebCore/style/values/fill-stroke/StyleStrokeWidth.h:
* Source/WebCore/style/values/primitives/StyleLengthWrapper+Blending.h:
* Source/WebCore/style/values/primitives/StyleLengthWrapper.h:
* Source/WebCore/style/values/primitives/StyleLengthWrapperData.h:
* Source/WebCore/style/values/primitives/StylePrimitiveNumeric.h:
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+Blending.h:
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+Evaluation.h:
* Source/WebCore/style/values/svg/StyleSVGStrokeDasharray.h:
* Source/WebCore/style/values/svg/StyleSVGStrokeDashoffset.h:

Canonical link: <a href="https://commits.webkit.org/300911@main">https://commits.webkit.org/300911@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c2c6504d2c856b05b99965fc3e8d15bddf1fcb37

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/124279 "11 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/43965 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/34689 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/131115 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/76349 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/126156 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/44708 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/52567 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/94543 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/76349 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/127233 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/35633 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/111168 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/75130 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/34577 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/29327 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/74596 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/105377 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/29550 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/133786 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/51191 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/39031 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/103019 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/51583 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/107386 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/102816 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26164 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/48178 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/26432 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/48131 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/51054 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/56836 "Built successfully") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/50492 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [❌ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/53848 "Failed to checkout and rebase branch from PR 51635") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/52167 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/rendering/style/RenderStyleInlines.h:
```diff
@@ -916,7 +913,7 @@ inline bool RenderStyle::insideSubmitButton() const { return m_rareInheritedData
 inline const Style::StrokeWidth& RenderStyle::strokeWidth() const { return m_rareInheritedData->strokeWidth; }
 inline bool RenderStyle::hasExplicitlySetStrokeWidth() const { return m_rareInheritedData->hasSetStrokeWidth; }
-inline bool RenderStyle::hasVisibleStroke() const { return hasStroke() && !strokeWidth().isZero(); }
+inline bool RenderStyle::hasVisibleStroke() const { return hasStroke() && strokeWidth().isPossiblyPositive(); }
```
Seems clear now that this function is subtly misnamed since it can return true in cases where the stroke ends up not being visible.

weinig committed:
[Style] Replace isZero/isNegative/isPositive functions inherited from WebCore::Length
https://bugs.webkit.org/show_bug.cgi?id=299972

Reviewed by Darin Adler.

`Style::LengthPercentage<>` and `Style::LengthWrapperBase<>` inherited (to help
things continue to work without too much churn) the isZero/isNegative/isPositive
functions from `WebCore::Length` and along with them their quirky behavior with
calc and keywords.

For a type that includes a Length, Percentage, Calc, and Keywords, the
functions have the following behavior:

- isZero:
    Length        !value
    Percentage    !value
    Calc          false
    Keyword       true (due to keywords having a value equal to 0)

- isPositive:
    Length        value > 0
    Percentage    value > 0
    Calc          true
    Keyword       false (due to keywords having a value equal to 0)

- isNegative:
    Length        value < 0
    Percentage    value < 0
    Calc          false
    Keyword       false (due to keywords having a value equal to 0)

This change replaces these functions with a set that hopefully have
more clear semantics:

The "known" value functions, `isKnownZero`, `isKnownPositive` and
`isKnownNegative`, will perform their comparison for Length and
Percentage values, but will return `false` for anything else (e.g. calc
or keywords).

The "possibly" value functions, `isPossiblyZero`, `isPossiblyPositive` and
`isPossiblyNegative`, will perform their comparison for Length and
Percentage values, but will return `true` for anything else (e.g. calc
or keywords).

In addition, the functions are constrained only to types where the types
numeric range makes the comparison meaningful. This means that a type
like `LengthPercentage<CSS::Nonnegative>`, the `isKnownNegative` and
`isPossiblyNegative` functions will unavailable and produce a compile
error if used. This allowed the removal of a number of useless calls.

* Source/WebCore/css/values/CSSValueConcepts.h:
* Source/WebCore/layout/formattingContexts/FormattingGeometry.cpp:
* Source/WebCore/layout/formattingContexts/block/BlockMarginCollapse.cpp:
* Source/WebCore/rendering/AutoTableLayout.cpp:
* Source/WebCore/rendering/GridLayoutFunctions.cpp:
* Source/WebCore/rendering/RenderBlock.cpp:
* Source/WebCore/rendering/RenderBlockFlow.cpp:
* Source/WebCore/rendering/RenderBox.cpp:
* Source/WebCore/rendering/RenderElement.cpp:
* Source/WebCore/rendering/RenderImage.cpp:
* Source/WebCore/rendering/RenderMarquee.cpp:
* Source/WebCore/rendering/RenderTable.cpp:
* Source/WebCore/rendering/RenderTableSection.cpp:
* Source/WebCore/rendering/cocoa/RenderThemeCocoa.mm:
* Source/WebCore/rendering/style/BorderData.cpp:
* Source/WebCore/rendering/style/BorderData.h:
* Source/WebCore/rendering/style/RenderStyle.cpp:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
* Source/WebCore/style/StyleExtractorCustom.h:
* Source/WebCore/style/values/StyleValueTypes.h:
* Source/WebCore/style/values/backgrounds/StyleBackgroundSize.h:
* Source/WebCore/style/values/backgrounds/StyleFillLayers.h:
* Source/WebCore/style/values/box/StyleMargin.h:
* Source/WebCore/style/values/fill-stroke/StyleStrokeWidth.cpp:
* Source/WebCore/style/values/fill-stroke/StyleStrokeWidth.h:
* Source/WebCore/style/values/primitives/StyleLengthWrapper+Blending.h:
* Source/WebCore/style/values/primitives/StyleLengthWrapper.h:
* Source/WebCore/style/values/primitives/StyleLengthWrapperData.h:
* Source/WebCore/style/values/primitives/StylePrimitiveNumeric.h:
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+Blending.h:
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+Evaluation.h:
* Source/WebCore/style/values/svg/StyleSVGStrokeDasharray.h:
* Source/WebCore/style/values/svg/StyleSVGStrokeDashoffset.h:

Canonical link: https://commits.webkit.org/300911@main
